### PR TITLE
add Spring Security

### PR DESCRIPTION
**Pull request by vremere-a**

### CODE REVIEW

File: src/main/java/com/dev/cinema/model/dto/hall/CinemaHallRequestDto.java [hunk 1]
```diff
@@ -1,9 +1,11 @@
 package com.dev.cinema.model.dto.hall;
 
+import javax.validation.constraints.NotNull;
 import lombok.Data;
 
 @Data
 public class CinemaHallRequestDto {
+    @NotNull
```
Comment (fyarkiy) [inline src/main/java/com/dev/cinema/model/dto/hall/CinemaHallRequestDto.java]: You should add message in bracket (message = "capacity shouldn't be null")

File: src/main/java/com/dev/cinema/validation/FieldsValueMatch.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    String message() default "Fields values don't match!";
+    String field();
+    String fieldMatch();
+    Class<?>[] groups() default {};
+    Class<? extends Payload>[] payload() default {};
+
+    @Target({ ElementType.TYPE })
+    @Retention(RetentionPolicy.RUNTIME)
+    @interface List {
+        FieldsValueMatch[] value();
```
Comment (fyarkiy) [inline src/main/java/com/dev/cinema/validation/FieldsValueMatch.java:23]: You don't need this block (20-23) - you can use it when you have several pairs to check (like password - repeatpassword, email - repeatEmail, etc)

**Review by fyarkiy — APPROVED** [review]

File: README.md [hunk 3]
```diff
@@ -1 +1,2 @@
-# cinema-project
+# CinemaTickets
+Cinema: Selling Tickets (Hibernate, Spring)
```
Comment (RomanDubovskyi) [inline README.md:2]: add description after you finish this project with single PR

File: src/main/java/com/dev/cinema/model/dto/hall/CinemaHallRequestDto.java (hunk 1 above)
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/model/dto/hall/CinemaHallRequestDto.java]: can int be null?

File: src/main/java/com/dev/cinema/model/dto/movie/MovieRequestDto.java [hunk 4]
```diff
@@ -1,9 +1,13 @@
 package com.dev.cinema.model.dto.movie;
 
+import javax.validation.constraints.NotNull;
+import javax.validation.constraints.Size;
 import lombok.Data;
 
 @Data
 public class MovieRequestDto {
+    @NotNull
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/model/dto/movie/MovieRequestDto.java]: ideally it's better to add some informative message

File: src/main/java/com/dev/cinema/model/dto/user/UserRegistrationDto.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
+package com.dev.cinema.model.dto.user;
+
+import com.dev.cinema.validation.EmailValidation;
+import com.dev.cinema.validation.FieldsValueMatch;
+import javax.validation.constraints.NotNull;
+import javax.validation.constraints.Size;
+import lombok.Data;
+
+@FieldsValueMatch.List({
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/model/dto/user/UserRegistrationDto.java]: do we need list here?

File: src/main/java/com/dev/cinema/validation/EmailValidation.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import javax.validation.Constraint;
+import javax.validation.Payload;
+
+@Documented
+@Constraint(validatedBy = EmailValidator.class)
+@Target({ElementType.METHOD, ElementType.FIELD })
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/validation/EmailValidation.java]: do we need to put this annotation on method ?

File: src/main/java/com/dev/cinema/validation/EmailValidator.java [hunk 7]
```diff
@@ -0,0 +1,17 @@
+package com.dev.cinema.validation;
+
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+
+public class EmailValidator implements ConstraintValidator<EmailValidation, String> {
+    private static final String REGEX = "^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+$";
+
+    @Override
+    public void initialize(EmailValidation constraintAnnotation) {
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/validation/EmailValidator.java]: it seems that we don't need this method

File: src/main/java/com/dev/cinema/validation/EmailValidator.java [hunk 8]
```diff
@@ -0,0 +1,17 @@
+package com.dev.cinema.validation;
+
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+
+public class EmailValidator implements ConstraintValidator<EmailValidation, String> {
+    private static final String REGEX = "^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+$";
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/validation/EmailValidator.java]:
```suggestion
    private static final String VALID_EMAIL_REGEX = "^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+$";
```
regex is not informative

File: src/main/java/com/dev/cinema/validation/FieldsValueMatchValidator.java [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    public void initialize(FieldsValueMatch constraintAnnotation) {
+        this.field = constraintAnnotation.field();
+        this.fieldMatch = constraintAnnotation.fieldMatch();
+    }
+
+    @Override
+    public boolean isValid(Object value, ConstraintValidatorContext context) {
+        Object fieldValue = new BeanWrapperImpl(value).getPropertyValue(field);
+        Object fieldMatchValue = new BeanWrapperImpl(value).getPropertyValue(fieldMatch);
+        return fieldValue != null ? fieldValue.equals(fieldMatchValue) : fieldMatchValue == null;
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/validation/FieldsValueMatchValidator.java]: you can use Objects.equals() here instead

**Review by RomanDubovskyi — APPROVED** [review]
fix comments and you can merge without second review

File: src/main/java/com/dev/cinema/model/dto/hall/CinemaHallRequestDto.java (hunk 1 above)
Comment (vremere-a) [reply src/main/java/com/dev/cinema/model/dto/hall/CinemaHallRequestDto.java]: done

Comment (vremere-a) [reply src/main/java/com/dev/cinema/model/dto/hall/CinemaHallRequestDto.java]: done

File: src/main/java/com/dev/cinema/model/dto/movie/MovieRequestDto.java (hunk 4 above)
Comment (vremere-a) [reply src/main/java/com/dev/cinema/model/dto/movie/MovieRequestDto.java]: done

File: src/main/java/com/dev/cinema/model/dto/user/UserRegistrationDto.java (hunk 5 above)
Comment (vremere-a) [reply src/main/java/com/dev/cinema/model/dto/user/UserRegistrationDto.java]: done

File: src/main/java/com/dev/cinema/validation/EmailValidation.java (hunk 6 above)
Comment (vremere-a) [reply src/main/java/com/dev/cinema/validation/EmailValidation.java]: done

File: src/main/java/com/dev/cinema/validation/EmailValidator.java (hunk 7 above)
Comment (vremere-a) [reply src/main/java/com/dev/cinema/validation/EmailValidator.java]: done

File: src/main/java/com/dev/cinema/validation/EmailValidator.java (hunk 8 above)
Comment (vremere-a) [reply src/main/java/com/dev/cinema/validation/EmailValidator.java]: done

File: src/main/java/com/dev/cinema/validation/FieldsValueMatchValidator.java (hunk 9 above)
Comment (vremere-a) [reply src/main/java/com/dev/cinema/validation/FieldsValueMatchValidator.java]: done

**Pull request merged**